add pdf as extension for rasters

dtedder opened the pull request.

dtedder committed:
allow the pdf extension and associate it with the raster layer type

dtedder committed:
adding layers from config file. connect raster layer load signal and remove if error or the layer has empty spatialReference

JamesMBallard commented on Shared/AddLocalDataController.cpp:
```diff
@@ -718,23 +719,44 @@ void AddLocalDataController::createFeatureLayerShapefile(const QString& path, in
+      if (auto* operationalLayers = ToolResourceProvider::instance()->operationalLayers(); operationalLayers)
+        operationalLayers->append(rasterLayer);
+
+    }, Qt::SingleShotConnection);
```
Nice usage of Qt::SingleShotConnection 👍.

JamesMBallard commented on Shared/LayerCacheManager.cpp:
```diff
@@ -442,7 +445,43 @@ void LayerCacheManager::connectSignals()
+
+            // check the spatial reference for the new layer for:
+            // - non-georeferenced PDFs may have been added
+            if (const auto spatialReference = rasterLayer->spatialReference(); spatialReference.isEmpty())
```
You don't need to hold onto the object at all here. 

```suggestion
            if (rasterLayer->spatialReference().isEmpty())
```

dtedder replied:
Good catch. Yep, not using this anywhere else below. Can't remember why I was storing a copy of this to begin with so I'll switch it out.

JamesMBallard commented on Shared/LayerCacheManager.cpp:
```diff
@@ -442,7 +445,43 @@ void LayerCacheManager::connectSignals()
+            {
+              operationalLayers->removeOne(rasterLayer);
+              rasterLayer->raster()->deleteLater();
+              rasterLayer->deleteLater();
```
Can you make one the parent of the other? That would avoid the multiple deleteLater calls on each one, but it may be more trouble than it's worth compared to being explicit like this.

Take this with a grain of salt. The suggestion may be worse.

dtedder replied:
This turned out to actually be a requirement. Removing layers is possible from the `[...]` menu in the ToC. Without setting the parent here was leading to memory growth that never came down. Have a look at this commit 743dce6c28e44dcd7bd54a03f8b4085250632dde. Adding and removing the same .dted file (one of our larger test rasters) over and over made it pretty clear. It may not have been noticeable before.

JamesMBallard reviewed (commented):
Looks pretty good. A few comments but not much to complain about here.

dtedder committed:
undo the previous change to the startup method that removes items from the list that do not have a valid spatial reference.
add red+italic style to layers loaded that have invalid spatial reference

dtedder committed:
minor cleanup of iterator and switch from int to qsizetype for a list count based loop

dtedder committed:
connect to the dataChanged signal from the operationalLayers to invalidate the UI when the spatialReference object is updated to a valid state for layers that load async

dtedder committed:
move back to using the proxy sorter as it was. moving the invalidate out of the proxy. force the ui to update on scene change to handle when scene packages are opened

dtedder committed:
put explicit back onto the constructor that takes single arg w default

dtedder committed:
use RasterLayer as the parent for the underlying Raster.
clean up layers as they are removed from the ToC

JamesMBallard commented on Shared/qml/TableOfContents.qml:
```diff
@@ -62,17 +62,21 @@ DsaPanel {
-                          toolController.alternateName(index)
-            onItemCheckedChanged: layerVisible = itemChecked
+            property bool spatialReferenceOk: toolController.spatialReferenceOk(index)
+            mainText: name && name !== "" ? name : toolController.alternateName(index)
```
presumably 'name' is a string, so it can't be null. It can be undefined but that seems uncommon to check for with strings. Do you think `name &&` is needed?

dtedder replied:
Didn't notice this when I adjusted the line. I'll remove the unnecessary null check.

JamesMBallard commented on Shared/AddLocalDataController.cpp:
```diff
@@ -718,23 +719,46 @@ void AddLocalDataController::createFeatureLayerShapefile(const QString& path, in
+
+      // check the spatial reference for the new layer for:
+      // - non-georeferenced PDFs may have been added
+      if (const auto spatialReference = rasterLayer->spatialReference(); spatialReference.isEmpty())
```
```suggestion
      if (rasterLayer->spatialReference().isEmpty())
```

JamesMBallard reviewed (approved):
Two more things, both minor. No need for another review but I am happy to review again if you want to change anything.

dtedder committed:
Update Shared/AddLocalDataController.cpp

Co-authored-by: James Ballard <JamesMBallard@users.noreply.github.com>

dtedder committed:
- removing unnecessary null check on QString
- switch to calling `deleteLater` instead of using `std::unique_ptr` to manage lifetime of layers removed from map

dtedder commented:
Thanks @JamesMBallard. I made another commit 65dd61b5a313794ceae2acb77c93920d47674728 switching to `deleteLater` instead of using `std::unique_ptr` based on your comment here: https://github.com/Esri/dynamic-situational-awareness-qt/pull/472#discussion_r2446019208.

May not be any harm. But as you suggest, probably better not to mix the two schemes. The layers are QObjects with a parent set. `deleteLater` seems more appropriate here too.
